### ISSUE TITLE
Server Start Up

### ISSUE THREAD

**Issue by ReStone** [ChimeraTK/ControlSystemAdapter-OPC-UA-Adapter#23]
There seem to be situations during server start up where OPC UA communication is started before adress space is fully populated.

In tests a client waiting for the server connected at once after starting the demo server but only some of the nodes have been working. Restarting the client fixed the problem.

The issue might also depend on system load. The worst scenario (many "missing" nodes) occured when starting the server application directly after starting the Ubuntu host - system was still loading.

**Comment by ichrispa** [comment]
Hi ReStone,

when the adapter is started, the server starts with it. Then an independant thread scans ChimeraTK for PVs and starts adding them. From what I gather from your issue, you have a hot-waiting client that immediately connects to the adapter. At that point in time, the adapter is still building up the lists.

In theory, this should not be a problem. If the client rescans the server/reconnects to it a couple of ms later, the nodes should exist. I can see however how some clients might just brake.

More technical:
in https://github.com/ChimeraTK/ControlSystemAdapter-OPC-UA-Adapter/blob/master/src/csa_opcua_adapter.cpp#L51:

```
csa_opcua_adapter::csa_opcua_adapter(boost::shared_ptr<ControlSystemPVManager> csManager, string configFile) {
	
	this->csManager = csManager; 	
	this->csa_opcua_adapter_InitServer(configFile);
	this->csa_opcua_adapter_InitVarMapping();
	
}
```

We need a server object to start the mapping, so just switching the lines would not work.

I would suggest the following solution:
  *  Do not start the Server - or the ipc_manager -  in csa_opcua_adapter_InitServer()
  * Call the ipc_manager->doStart after initVarMapping in the constructor.

This would delay any network I/O till after the variable mapping is complete.

We're on it, I'll notify you when a new version is pushed.

Best regards,

ichrispa

**Comment by ReStone** [comment]
Perfect!

**Comment by jmrahm** [comment]
Hey ReStone,

we try to fix it. Now the server starts after all variables are mapped. Please, try it again. ;-)

Best regards
jmrahm

**Comment by ReStone** [comment]
Finally I had some time to test the new version. It seems to work fine now! In all of several server startups, the client connection (WinCC) "woke up" with all configured nodes. Thank you for the fix!

One strange thing: the new version starts with many error messages of following kind:
![startup-errors](https://cloud.githubusercontent.com/assets/19655759/25993094/2ec7b4a4-3709-11e7-9d7d-652fb5cf0c5c.PNG)
What does that mean?

**Comment by ichrispa** [comment]
Hi ReStone,

since 0.2, we have implemented a lot of model checking procedures in the stack. There are several conditions concerning when an object/type may be referenced by another object. We add 1544 nodes from namespace 0 as a base model to the stack as a common ground for communications.

The errors you are seeing basically result because some nodes do not conform to the requirements and the stack refuses to add them. They are inconsequential to your server, because we only use very rudimentary models. Chances are that for example some attribute nodes of the StateType could not be added correctly, which will never affect your server model because we don't use them.

The errors are quite annoying and there was a lengthy discussion about not including these checks in 0.2. But the checks are formally correct; we just have some problems with the initialization sequence of namespace 0.

In short: The errors are annoying, but have no functional effect on your application as far as I can see.

Best regards,

Chris

**Comment by ichrispa** [comment]
If the errors are a problem, we can "scale down" the namespace 0 to its essentials; that would limit future expansion to more complex models, but the errors would be gone...

**Comment by ReStone** [comment]
The "scale down" should definitely be done for a final adapter version. Momentarily I think we can live with it. Thanks for the clarification!
